### PR TITLE
[#39] Optimize the workflow to generate the "sample" project before executing CI actions

### LINKED ISSUE #39: Optimize the workflow to generate the "sample" project before executing CI actions
## Why

The step to generate the working project to run tests takes a significant time to complete `> 3 mins`.

image

The need to generate a sample project first to execute the test actions in Android and iOS platforms, and to push the sample project to the repository 👉 duplicating the generation step 3 times is not an ideal approach.

We must improve the test workflow to **generate the sample project before running test actions on Android or iOS platforms**.

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by luongvo**
- Close #39

## What happened 👀

- Modify the `test.yml` workflow to generate the sample project before executing test actions on Android or iOS platforms.
- Add cancel previous runs.
- Separate danger report between Android and iOS.
- Fix: Running Danger SwiftFormat from the non-root directory.
- Add the initial `contribution guide` for git-submodule.

## Insight 📝

The Danger for iOS has an issue when running from the non-root directory: `sample/ios`, which makes the SwiftFormat can not file expected files because of duplicating the `sample/ios` path https://github.com/nimblehq/kmm-templates/actions/runs/5982564343/job/16231801560 💥 

image

- The cause is while [`danger-ruby-swiftformat` returns the list of swift files with the full path from the root dir](https://github.com/garriguv/danger-ruby-swiftformat/blob/6c7828baff849d03665bce511f141e59952bba96/lib/swiftformat/plugin.rb#L81-L96), e.g. `sample/ios/fastlane/Constants/Constant.swift`, including `sample/ios`, `SwiftFormat` joins them with the current working dir `/Users/runner/work/kmm-templates/kmm-templates/sample/ios/`.
- There is no official support to bypass this issue. I made a workaround to modify the `danger-ruby-swiftformat` source to remove the `sample/ios` prefix from all files by using this operator `.map { |file| file.partition('sample/ios/').last }` 🙏 

	```
      files
        .select { |file| file.end_with?(".swift") }
        .reject { |file| @exclude.any? { |glob| File.fnmatch(glob, file) } }
        .uniq.map { |file| file.partition('sample/ios/').last }
        .sort
	```

## Proof Of Work 📹

The workflow works properly as it is.

### CODE REVIEW

**Comment by luongvo** [comment]
@phongvhd93 @markgravity @kaungkhantsoe This PR is now ready for review, too 🎉

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -1,26 +1,75 @@
...
     steps:
-      - name: Check out source code
+      - name: Check out
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 100
+          submodules: "recursive"
 
-      - name: Set up JDK
+      - name: Set up Java JDK
```
Comment (phongvhd93) [inline .github/workflows/test.yml]: the J in JDK stands for Java, is that right? so no need JAVA here

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -73,20 +129,12 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-gems-
 
-      - name: Setup Java JDK
-        uses: actions/setup-java@v2.1.0
+      - name: Set up Java JDK
```
Comment (phongvhd93) [inline .github/workflows/test.yml]: J is Java already 🤔 so no need to mention Java

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -62,7 +117,8 @@ jobs:
       - name: Check out
         uses: actions/checkout@v3
         with:
-          submodules: recursive
+          fetch-depth: 100
```
Comment (phongvhd93) [inline .github/workflows/test.yml]: Why do we have to set the fetch depth = 100 here? I think shallow clone with depth = 1 is enough.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -1,26 +1,75 @@
...
     timeout-minutes: 30
+    defaults:
+      run:
+        working-directory: ./sample
     steps:
-      - name: Check out source code
+      - name: Check out
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 100
```
Comment (phongvhd93) [inline .github/workflows/test.yml]: if you agree with reducing fetch depth, don't forget to update here 🙏

File: .github/workflows/test.yml (hunk 3 above)
Comment (luongvo) [reply .github/workflows/test.yml]: Fixed in https://github.com/nimblehq/kmm-templates/pull/38/commits/4845f136d9c3489313ade13983ca208df75a1c17

**Review by phongvhd93 — APPROVED** [review]

**Review by kaungkhantsoe — APPROVED** [review]
LGTM 👍

**Pull request merged**